### ISSUE TITLE
how can I reconstruct images (MAE)

### ISSUE THREAD

**Issue by DanTaranis** [lightly-ai/lightly#991]
Hi and thank you for writing this great library.

I'm trying to figure out if my MAE training worked beyond looking at mse numbers. 
I input an image of shape 256x256 (rgb) and then I get predictions of shape bx38x3072
I'm using the default torchvision.models.vit_b_32

I am not sure how to reshape this back into an image? it also just doesn't make sense based on the numbers - there should be 64 patches of 768 (=32*32*3)?

**Comment by DanTaranis** [comment]
I figure out the issue - the default vit.sequence length is 50 , since I used 256 instead of 224 I should have adjusted the length

**Comment by guarin** [comment]
Hi, converting predictions back to images is indeed not straightforward. You are correct that you have to adapt the sequence length when changing the input image or patch size.

Regarding transforming the tokens back to images you could try using the `unpatchify` method from the paper: https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/models_mae.py#L109

I didn't try it on outputs from our MAE yet so maybe it does not work perfectly. Please tell us if you get it working :)

**Comment by DanTaranis** [comment]
thanks. I'll try it out!

**Comment by DanTaranis** [comment]
kinda hood but:
```
    def predict(self, images):
        batch_size = images.shape[0]
        idx_keep, idx_mask = utils.random_token_mask(
            size=(batch_size, self.sequence_length),
            mask_ratio=self.mask_ratio,
            device=images.device,
        )
        x_encoded = self.forward_encoder(images, idx_keep)
        x_pred = self.forward_decoder(x_encoded, idx_keep, idx_mask)

        # get image patches for masked tokens
        patches = utils.patchify(images, self.patch_size)
        # must adjust idx_mask for missing class token
        target = utils.get_at_index(patches, idx_mask - 1)

        #reshape back into image
        x_pred = x_pred.reshape(shape=(target.shape[0], target.shape[1], self.patch_size, self.patch_size, 3))
        target = target.reshape(shape=(target.shape[0], target.shape[1], self.patch_size, self.patch_size, 3))

        x_pred_images = torch.zeros(size=(x_pred.shape[0],images.shape[2],images.shape[3],images.shape[1]))
        target_images = torch.zeros(size=(x_pred.shape[0],images.shape[2],images.shape[3],images.shape[1]))
        for idx, orginal_idx in enumerate(idx_mask[0]):
            orginal_idx -= 1
            i, j = orginal_idx//(images.shape[2]/self.patch_size), orginal_idx%(images.shape[3]/self.patch_size)
            x_pred_images[:,int(i*self.patch_size):int((i+1)*self.patch_size), int((j)*self.patch_size):int((j+1)*self.patch_size), :] = x_pred[:,idx,:, :,:]
            target_images[:,int(i*self.patch_size):int((i+1)*self.patch_size), int((j)*self.patch_size):int((j+1)*self.patch_size), :] = target[:,idx,:, :,:]

        return x_pred_images, target_images
        ```

**Comment by guarin** [comment]
Nice! If you just want to run inference you can also drop the whole idx business and get the predictions for the full image like this:

```
    def predict(self, images):
        x_encoded = self.forward_encoder(images)
        x_pred = self.forward_decoder(x_encoded)
        x_pred = x_pred[:, 1:] # drop class token
        ...
```

**Comment by DanTaranis** [comment]
wouldn't it give the model too much information? I want to see how it does when part of the image is masked...

**Comment by guarin** [comment]
Yes, you are right :) I never tested this, I am not sure after how much training the model can recreate the input perfectly without any masking.

**Comment by siavashk** [comment]
This is not correct. I spent some time and used the `unpatchify` function from 8878b5b5e816bbd1f11b5c0a866eaee45a9a6004 and the following works:
```
def combine_masked_unmasked(masked_tensor, unmasked_tensor, masked_indices, unmasked_indices):
    batch_size, num_masked_tokens, patch_size = masked_tensor.shape
    num_unmasked_tokens = unmasked_tensor.shape[1]
    
    # Calculate total number of tokens
    total_tokens = num_masked_tokens + num_unmasked_tokens
    
    # Initialize the output tensor
    output_tensor = torch.zeros((batch_size, total_tokens, patch_size), dtype=masked_tensor.dtype)
    
    for i in range(batch_size):
        # Combine masked and unmasked indices
        combined_indices = torch.cat((masked_indices[i], unmasked_indices[i]), dim=0)
        # Create a list of (index, source_tensor, source_index)
        indexed_sources = []
        for j in range(num_masked_tokens):
            indexed_sources.append((masked_indices[i, j].item(), masked_tensor[i, j, :]))
        for j in range(num_unmasked_tokens):
            indexed_sources.append((unmasked_indices[i, j].item(), unmasked_tensor[i, j, :]))
        
        # Sort based on the original positions
        indexed_sources.sort(key=lambda x: x[0])
        
        # Fill the output tensor
        for j, (_, patch) in enumerate(indexed_sources):
            output_tensor[i, j, :] = patch
    
    return output_tensor
```
You need to modify the `forward` pass of the `MAE` class to the following:
```
def forward(self, images):
    batch_size = images.shape[0]
    idx_keep, idx_mask = utils.random_token_mask(
        size=(batch_size, self.sequence_length),
        mask_ratio=self.mask_ratio,
        device=images.device,
    )
    x_encoded = self.forward_encoder(images=images, idx_keep=idx_keep)
    x_pred = self.forward_decoder(
        x_encoded=x_encoded, idx_keep=idx_keep, idx_mask=idx_mask
    )

    # get image patches for masked tokens
    patches = utils.patchify(images, self.patch_size)
    # must adjust idx_mask for missing class token
    target = utils.get_at_index(patches, idx_mask - 1)
    
    unmasked_target = utils.get_at_index(patches, idx_keep[:, 1:] - 1)
    reconstructed = combine_masked_unmasked(x_pred, unmasked_target, idx_mask - 1, idx_keep[:, 1:] - 1)
    reconstructed = unpatchify(reconstructed, self.patch_size, 1).to(x_pred.device)
    
    return x_pred, target, reconstructed
```

Some explanation about the logic:

1. If your image is divided into `N` patches, there are going to be `N+1` tokens. This is because the first token is the `CLS_TOKEN` in vision transformers. This is why I pass `idx_mask - 1` to the `combine_masked_unmasked` function.
2. The `CLS_TOKEN` is always at index 0.
3. The unmasked tokens always start with 0 which corresponds to the `CLS_TOKEN`. This is why the tensor is sliced with `idx_keep[:, 1:]`.

The `reconstructed` batch is of shape `batch_size x n_channels x height x width`.

**Comment by goncayilmaz** [comment]
Thanks, this is really helpful!
